solr/core/src/java/org/apache/solr/metrics/prometheus/SolrPrometheusRegistry.java

```diff
@@ -0,0 +1,153 @@
+import io.prometheus.metrics.model.registry.PrometheusRegistry;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public abstract class SolrPrometheusRegistry {
+  PrometheusRegistry prometheusRegistry;
+  String registryName;
+  private final Map<String, Counter> metricCounters;
+  private final Map<String, Gauge> metricGauges;
+
+  public SolrPrometheusRegistry(PrometheusRegistry prometheusRegistry) {
+    this.prometheusRegistry = prometheusRegistry;
+    this.metricCounters = new HashMap<>();
+    this.metricGauges = new HashMap<>();
+  }
+
+  public PrometheusRegistry getPrometheusRegistry() {
+    return prometheusRegistry;
+  }
+
+  public String getRegistryName() {
+    return registryName;
+  }
+
+  public Map<String, Counter> copyOfCounters() {
+    return ImmutableMap.copyOf(metricCounters);
+  }
+
+  public Map<String, Gauge> copyOfGauges() {
+    return ImmutableMap.copyOf(metricGauges);
+  }
+
+  public void exportMeter(
+      Meter dropwizardMetric, String prometheusMetricName, Map<String, String> labelsMap) {
+    if (!copyOfCounters().containsKey(prometheusMetricName)) {
+      ArrayList<String> labels = new ArrayList<>(labelsMap.keySet());
+      registerCounter(prometheusMetricName, labels.toArray(String[]::new));
+    }
+    ArrayList<String> labelValues = new ArrayList<>(labelsMap.values());
+    getMetricCounter(prometheusMetricName)
+        .labelValues(labelValues.toArray(String[]::new))
+        .inc(dropwizardMetric.getCount());
+  }
+
+  public void exportCounter(
+      com.codahale.metrics.Counter dropwizardMetric,
+      String prometheusMetricName,
+      Map<String, String> labelsMap) {
+    if (!copyOfCounters().containsKey(prometheusMetricName)) {
+      ArrayList<String> labels = new ArrayList<>(labelsMap.keySet());
+      registerCounter(prometheusMetricName, labels.toArray(String[]::new));
+    }
+    ArrayList<String> labelValues = new ArrayList<>(labelsMap.values());
+    getMetricCounter(prometheusMetricName)
+        .labelValues(labelValues.toArray(String[]::new))
+        .inc(dropwizardMetric.getCount());
+  }
+
+  public void exportTimer(
+      Timer dropwizardMetric, String prometheusMetricName, Map<String, String> labelsMap) {
+    if (!copyOfGauges().containsKey(prometheusMetricName)) {
+      ArrayList<String> labels = new ArrayList<>(labelsMap.keySet());
+      registerGauge(prometheusMetricName, labels.toArray(String[]::new));
+    }
+    ArrayList<String> labelValues = new ArrayList<>(labelsMap.values());
+    getMetricGauge(prometheusMetricName)
+        .labelValues(labelValues.toArray(String[]::new))
+        .set(dropwizardMetric.getMeanRate());
+  }
+
+  public void exportGauge(
+      com.codahale.metrics.Gauge<?> dropwizardMetricRaw,
+      String prometheusMetricName,
+      Map<String, String> labelsMap) {
+    Object dropwizardMetric = (dropwizardMetricRaw).getValue();
+    if (!copyOfGauges().containsKey(prometheusMetricName)) {
```

Comment:
Why use `copyOfGauges()` in the check if it doesn't seem to be referenced again?

Reply:
Good point. Going to remove this and just reference the map. Immutable map isn't allowed upstream anyways and breaks tests.